pallets/gear-builtin-actor/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! # Gear Builtin Actors Pallet
+//!
+//! The Builtn Actors pallet provides a registry of the builtin actors available in the Runtime.
+//!
+//! - [`Config`]
+//!
+//! ## Overview
+//!
+//! The pallet implements the `pallet_gear::BuiltinRouter` allowing to restore builtin actors
+//! claimed `BuiltinId`'s based on their corresponding `ProgramId` address.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::items_after_test_module)]
+
+extern crate alloc;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking;
+
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub use weights::WeightInfo;
+
+use alloc::{collections::BTreeMap, string::ToString};
+use core_processor::{
+    common::{DispatchOutcome, JournalNote},
+    process_non_executable, SystemReservationContext,
+};
+use gear_core::{
+    ids::{BuiltinId, MessageId, ProgramId},
+    message::{DispatchKind, ReplyMessage, ReplyPacket, StoredDispatch},
+};
+use gear_core_errors::SimpleExecutionError;
+use impl_trait_for_tuples::impl_for_tuples;
+use pallet_gear::{BuiltinRouter, BuiltinRouterProvider};
+use parity_scale_codec::{Decode, Encode};
+use sp_io::hashing::blake2_256;
+use sp_runtime::traits::{TrailingZeroInput, Zero};
+use sp_std::prelude::*;
+
+pub use pallet::*;
+
+#[allow(dead_code)]
+const LOG_TARGET: &str = "gear::builtin_actor";
+
+pub trait Dispatchable {
+    type Payload: AsRef<[u8]>;
+
+    fn source(&self) -> ProgramId;
+    fn destination(&self) -> BuiltinId;
+    fn payload_bytes(&self) -> &[u8];
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct SimpleBuiltinMessage {
```

Comment:
you can avoid this type by operating some config asoc: Dispatchable

Reply:
as well as impl from into to message type used in send_dispatch ext_manager

Reply:
that makes sense.